pallets/parachain-staking/src/try_state.rs

```diff
@@ -0,0 +1,202 @@
+// KILT Blockchain – https://botlabs.org
+// Copyright (C) 2019-2023 BOTLabs GmbH
+
+// The KILT Blockchain is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The KILT Blockchain is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// If you feel like getting in touch with us, you can do so at info@botlabs.org
+
+use frame_support::{ensure, traits::Get};
+use kilt_support::test::convert_error_message;
+use scale_info::prelude::format;
+use sp_runtime::{traits::Zero, SaturatedConversion, Saturating};
+
+use crate::{
+	set::OrderedSet,
+	types::{BalanceOf, Candidate, Stake},
+	CandidatePool, Config, DelegatorState, LastDelegation, MaxCollatorCandidateStake, MaxSelectedCandidates, Pallet,
+	Round, TopCandidates, TotalCollatorStake,
+};
+
+pub(crate) fn do_try_state<T: Config>() -> Result<(), &'static str> {
+	validate_candiate_pool::<T>()?;
+	validate_top_candidates::<T>()?;
+	validate_stake::<T>()
+}
+
+fn validate_candiate_pool<T: Config>() -> Result<(), &'static str> {
+	// check if enough collators are set.
+	ensure!(
+		CandidatePool::<T>::count() >= T::MinCollators::get(),
+		convert_error_message(format!(
+			"Insufficient collators. Collators count: {:?}. Min required collators: {:?}",
+			CandidatePool::<T>::count(),
+			T::MinCollators::get()
+		))
+	);
+
+	CandidatePool::<T>::iter_values().try_for_each(
+		|candidate: Candidate<T::AccountId, BalanceOf<T>, _>| -> Result<(), &'static str> {
+			let sum_delegations: BalanceOf<T> = candidate
+				.delegators
+				.iter()
+				.fold(Zero::zero(), |acc, stake| acc.saturating_add(stake.amount));
+
+			// total stake should be the sum of delegators stake + colator stake.
+			let stake_total = sum_delegations.saturating_add(candidate.stake);
+			ensure!(
+				stake_total == candidate.total,
+				convert_error_message(format!("Total stake of collator {:?} does not match", candidate.id))
```

Comment:
Should we add what the expected and returned values are? In general if we are testing a condition, we might want to also want to output, in the case of a failure, what the expected condition was and what was actually returned.

Reply:
I updated all possible error messages